compiler/src/main/java/com/squareup/anvil/compiler/codegen/MergeAnnotationsCheckGenerator.kt

```diff
@@ -0,0 +1,76 @@
+package com.squareup.anvil.compiler.codegen
+
+import com.google.auto.service.AutoService
+import com.squareup.anvil.compiler.api.AnvilContext
+import com.squareup.anvil.compiler.api.CodeGenerator
+import com.squareup.anvil.compiler.internal.reference.AnnotationReference
+import com.squareup.anvil.compiler.internal.reference.AnvilCompilationExceptionClassReference
+import com.squareup.anvil.compiler.internal.reference.classAndInnerClassReferences
+import com.squareup.anvil.compiler.mergeComponentFqName
+import com.squareup.anvil.compiler.mergeInterfacesFqName
+import com.squareup.anvil.compiler.mergeModulesFqName
+import com.squareup.anvil.compiler.mergeSubcomponentFqName
+import org.jetbrains.kotlin.descriptors.ModuleDescriptor
+import org.jetbrains.kotlin.psi.KtFile
+import java.io.File
+
+@AutoService(CodeGenerator::class)
+internal class MergeAnnotationsCheckGenerator : PrivateCodeGenerator() {
+  override fun isApplicable(context: AnvilContext): Boolean = !context.disableComponentMerging
+
+  override fun generateCodePrivate(
+    codeGenDir: File,
+    module: ModuleDescriptor,
+    projectFiles: Collection<KtFile>
+  ) {
+    projectFiles
+      .classAndInnerClassReferences(module)
+      .forEach { clazz ->
+        val annotations = clazz.annotations
+          .findAll(
+            mergeComponentFqName,
+            mergeSubcomponentFqName,
+            mergeModulesFqName,
+            mergeInterfacesFqName
+          )
+          .ifEmpty { return@forEach }
+
+        annotations.checkSingleAnnotation()
+        annotations.checkNoDuplicateScope(contributeAnnotation = false)
```

Comment:
I had to duplicate these checks between the `InterfacerMerger`, `ModuleMerger` and `ModuleMergerIr`. That was annoying. Instead, it was a lot easier to move these checks into the compiler frontend with its own code generator.